logic/resolvable/Concludable.java

```diff
@@ -268,67 +267,61 @@ public Relation(final RelationConstraint constraint, @Nullable IsaConstraint isa
-
-            return matchRolePlayerIndices(conjRolePlayers, thenRolePlayers, new HashMap<>(), conceptMgr)
-                    .map(indexMap -> rolePlayerMappingToUnifier(indexMap, thenRolePlayers, unifierBuilder.duplicate(), conceptMgr));
-        }
-
-        private ResourceIterator<Map<RolePlayer, Set<Integer>>> matchRolePlayerIndices(
-                List<RolePlayer> conjRolePlayers, List<RolePlayer> thenRolePlayers,
-                Map<RolePlayer, Set<Integer>> mapping, ConceptManager conceptMgr) {
-
-            if (conjRolePlayers.isEmpty()) return Iterators.iterate(list(mapping));
-            RolePlayer conjRP = conjRolePlayers.get(0);
-
-            return Iterators.iterate(IntStream.range(0, thenRolePlayers.size()).iterator())
-                    .filter(thenIdx -> mapping.values().stream().noneMatch(players -> players.contains(thenIdx)))
-                    .filter(thenIdx -> unificationSatisfiable(conjRP, thenRolePlayers.get(thenIdx), conceptMgr))
-                    .map(thenIdx -> {
-                        Map<RolePlayer, Set<Integer>> clone = cloneMapping(mapping);
+            Set<RolePlayer> thenRolePlayers = relationConclusion.relation().players();
+
+            return matchRolePlayers(conjRolePlayers, thenRolePlayers, new HashMap<>(), conceptMgr)
+                    .map(mapping -> convertRPMappingToUnifier(mapping, unifierBuilder.duplicate(), conceptMgr));
+        }
+
+        private ResourceIterator<Map<RolePlayer, Set<RolePlayer>>> matchRolePlayers(
+                List<RolePlayer> conjRolePLayers, Set<RolePlayer> thenRolePlayers,
+                Map<RolePlayer, Set<RolePlayer>> mapping, ConceptManager conceptMgr) {
+            if (conjRolePLayers.isEmpty()) return Iterators.iterate(list(mapping));
+            RolePlayer conjRP = conjRolePLayers.get(0);
+            return Iterators.iterate(thenRolePlayers)
+                    .filter(thenRP -> Iterators.iterate(mapping.values()).noneMatch(rolePlayers -> rolePlayers.contains(thenRP)))
+                    .filter(thenRP -> unificationSatisfiable(conjRP, thenRP, conceptMgr))
+                    .map(thenRP -> {
+                        Map<RolePlayer, Set<RolePlayer>> clone = cloneMapping(mapping);
                         clone.putIfAbsent(conjRP, new HashSet<>());
-                        clone.get(conjRP).add(thenIdx);
+                        clone.get(conjRP).add(thenRP);
                         return clone;
-                    }).flatMap(newMapping -> matchRolePlayerIndices(conjRolePlayers.subList(1, conjRolePlayers.size()),
-                                                                    thenRolePlayers, newMapping, conceptMgr));
+                    }).flatMap(newMapping -> matchRolePlayers(conjRolePLayers.subList(1, conjRolePLayers.size()),
+                                                              thenRolePlayers, newMapping, conceptMgr));
+        }
+
+        private Unifier convertRPMappingToUnifier(
```

Comment:
We still need to create a mapping from `RolePlayer` to `RolePlayer`, and then turn that into a unification. I looked into doing this all in a single loop, but there is no way to know if a Then RolePlayer has been seen before. I would have to change the `Unification.Builder` object to do that. The current solution therefore seems more intuitive.